staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -600,58 +599,56 @@ func (f *DeltaFIFO) Replace(list []interface{}, _ string) error {
 		}
 	}
 
-	if f.knownObjects == nil {
-		// Do deletion detection against our own list.
-		queuedDeletions := 0
-		for k, oldItem := range f.items {
-			if keys.Has(k) {
-				continue
-			}
-			// Delete pre-existing items not in the new list.
-			// This could happen if watch deletion event was missed while
-			// disconnected from apiserver.
-			var deletedObj interface{}
-			if n := oldItem.Newest(); n != nil {
-				deletedObj = n.Object
-			}
-			queuedDeletions++
-			if err := f.queueActionLocked(Deleted, DeletedFinalStateUnknown{k, deletedObj}); err != nil {
-				return err
-			}
-		}
-
-		if !f.populated {
-			f.populated = true
-			// While there shouldn't be any queued deletions in the initial
-			// population of the queue, it's better to be on the safe side.
-			f.initialPopulationCount = keys.Len() + queuedDeletions
-		}
-
-		return nil
-	}
-
-	// Detect deletions not already in the queue.
-	knownKeys := f.knownObjects.ListKeys()
+	// Do deletion detection against objects in the queue
 	queuedDeletions := 0
-	for _, k := range knownKeys {
+	for k, oldItem := range f.items {
 		if keys.Has(k) {
 			continue
 		}
-
-		deletedObj, exists, err := f.knownObjects.GetByKey(k)
-		if err != nil {
-			deletedObj = nil
-			klog.Errorf("Unexpected error %v during lookup of key %v, placing DeleteFinalStateUnknown marker without object", err, k)
-		} else if !exists {
-			deletedObj = nil
-			klog.Infof("Key %v does not exist in known objects store, placing DeleteFinalStateUnknown marker without object", k)
+		// Delete pre-existing items not in the new list.
+		// This could happen if watch deletion event was missed while
+		// disconnected from apiserver.
+		var deletedObj interface{}
+		if n := oldItem.Newest(); n != nil {
+			deletedObj = n.Object
+
+			// if the previous object is a DeletedFinalStateUnknown, we have to extract the actual Object
+			if d, ok := deletedObj.(DeletedFinalStateUnknown); ok {
+				deletedObj = d.Obj
+			}
 		}
 		queuedDeletions++
 		if err := f.queueActionLocked(Deleted, DeletedFinalStateUnknown{k, deletedObj}); err != nil {
 			return err
 		}
 	}
 
+	if f.knownObjects != nil {
```

Comment:
I will read the comments in a second but I'm very confused about why you moved this block and switched its condition?

Reply:
ok -- read the comments and I'm still confused, can you put it back to minimize the diff or document why it has to change?

Reply:
👋 Its mostly to get consistent behavior for when the object is present in `KnownObjects` or not when creating the deleted event with `DeletedFinalStateUnknown`.

Did you read the commit message? I'm happy to drop this and we can make a PR, or make it more clear via comments in the code.
https://github.com/kubernetes/kubernetes/pull/115620/commits/7bcc3e00fc28b2548886d04639a2e352ab37fb55


Also, this test change shows the actual "behavior change" https://github.com/kubernetes/kubernetes/pull/115620/commits/7bcc3e00fc28b2548886d04639a2e352ab37fb55#diff-b1aad36a628c7aeab74b3761643a965d90c258d019c4f77645ff45c51eb6ca35R498, as well as the rest of the test additions in `TestDeltaFIFO_ReplaceMakesDeletions`.

Reply:
ok, the test might have won me over, I will read this again in the morning. :)

Reply:
it makes sense to me now.